### PR TITLE
Adding a unit test to check resource's data

### LINKED ISSUE #1449: Create unit test to ensure resources structure matches its directory location
There have been cases in development where you are working on a resource inside a given namespace's but forget to update the namespace setting inside the yaml.
Write a unit test to validate each resource is in the directory it should be, i.e. validate namespace, kind, etc.

### PR DESCRIPTION
**Pull request by thekad**
To avoid possibly checking in or working with yaml files that don't
match their filesystem distribution, example: working in a file that is
checked-in to a particular namespace but the yaml says default.
Specifically walking the data directory instead of polling `AssetNames()`
so running bindata isn't necessary to run this test.

```release-note
NONE
```
Closes #1449

### CODE REVIEW

File: pkg/sync/v5/translate_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		o, err := unmarshal(b)
+		if err != nil {
+			return fmt.Errorf("error unmarshaling %s: %v", path, err)
+		}
+
+		parts := strings.Split(path, "/")
+		kind := parts[1]
+		// cluster-wide kinds have an empty string as namespace
+		namespace := ""
+		if !clusterKinds[kind] {
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: if len(parts) == 3

File: pkg/sync/v5/translate_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+			namespace = parts[2]
+		}
+
+		// check the namespace matches
+		if namespace != o.GetNamespace() {
+			return fmt.Errorf("%s namespace doesn't match its path (%s != %s)", path, namespace, o.GetNamespace())
+		}
+
+		// check the kind is the prefix for the given template
+		prefix := o.GetKind()
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: can also partially check the apiVersion

File: pkg/sync/v5/translate_test.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		kind := parts[1]
+		// cluster-wide kinds have an empty string as namespace
+		namespace := ""
+		if !clusterKinds[kind] {
+			namespace = parts[2]
+		}
+
+		// check the namespace matches
+		if namespace != o.GetNamespace() {
+			return fmt.Errorf("%s namespace doesn't match its path (%s != %s)", path, namespace, o.GetNamespace())
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: gather a []error and print it at the end

File: pkg/sync/v5/translate_test.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+func TestTranslatableData(t *testing.T) {
+	clusterKinds := map[string]bool{
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: can remove this

File: pkg/sync/v5/translate_test.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		oKind := o.GetKind()
+		if strings.Contains(apiVersion, "/") {
+			oKind = fmt.Sprintf("%s.%s", oKind, strings.Split(apiVersion, "/")[0])
+		}
+		if !strings.HasPrefix(kind, oKind) {
+			errors = append(errors, fmt.Sprintf("%s kind doesn't match its path (%s != %s)", path, kind, oKind))
+		}
+
+		// check the name matches, caveat: all colons (":") are replaced with single dash ("-")
+		oName := re.ReplaceAllString(o.GetName(), "-")
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: Avoid using regular expressions whenever you can.  Replace with `oName := strings.Replace(o.GetName(), ":", "-", -1)`

File: pkg/sync/v5/translate_test.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+			errors = append(errors, fmt.Sprintf("%s namespace doesn't match its path (%s != %s)", path, namespace, o.GetNamespace()))
+		}
+
+		// check the kind is the prefix for the given template
+		apiVersion := o.GetAPIVersion()
+		oKind := o.GetKind()
+		if strings.Contains(apiVersion, "/") {
+			oKind = fmt.Sprintf("%s.%s", oKind, strings.Split(apiVersion, "/")[0])
+		}
+		if !strings.HasPrefix(kind, oKind) {
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: why can't this be `if kind != oKind` ?

File: pkg/sync/v5/translate_test.go (hunk 5 above)
Comment (thekad) [reply pkg/sync/v5/translate_test.go]: Ah but there are some names that have e.g. `system::foo:bar` (double colon) and those are collapsed into a single dash. Either I find a way to collapse one or more `:` into a single dash (which proved snap easy with regexp) or I rename the files to e.g. `system--foo-bar.yaml`

Comment (jim-minter) [reply pkg/sync/v5/translate_test.go]: Ha, that's a discrepancy I hadn't spotted.  Please use strings.Replace() and rename the files.

**Comment by jim-minter** [comment]
Also please squash.

File: pkg/sync/v5/translate_test.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		apiVersion := o.GetAPIVersion()
+		oKind := o.GetKind()
+		if strings.Contains(apiVersion, "/") {
+			oKind = fmt.Sprintf("%s.%s", oKind, strings.Split(apiVersion, "/")[0])
+		}
+		if kind != oKind {
+			errors = append(errors, fmt.Sprintf("%s kind doesn't match its path (%s != %s)", path, kind, oKind))
+		}
+
+		// check the name matches, caveat: all colons (":") are replaced with single dash ("-")
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: comment no longer fits

File: pkg/sync/v5/translate_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+func TestTranslatableData(t *testing.T) {
+	var errors []string
+
+	err := filepath.Walk("data", func(path string, info os.FileInfo, errIn error) error {
+		if errIn != nil {
+			return errIn
+		}
+
+		if info.IsDir() || !strings.HasSuffix(info.Name(), ".yaml") {
```
Comment (jim-minter) [inline pkg/sync/v5/translate_test.go]: prob should remove the .yaml suffix check and not allow any other files in there

**Pull request merged**